### PR TITLE
[Docs] Storage docs rewrite

### LINKED ISSUE #3123: [Docs] Storage docs rewrite
Storage docs need to be more focused on the jobs-to-be-done. 

Creating an issue to track for 0.5.

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Closes #3123.

TODO:
- [x] Mounting mode considerations section - a nice table to help users pick b/w `COPY` and `MOUNT`

Tested:
- [x] Rendered locally

### CODE REVIEW

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:3]: Cloud Object Storage / Cloud Buckets?

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:1]: Note "object storage" is the more common term than "object stores". Replace throughout?

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:9]:
```suggestion
Buckets are made available to each task at a local path on the remote VM, so
the task can access bucket objects as if they were local files.
```

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:92]: This is excellent! Very clear. 

Some first impression confusion
- `my_data` -> `mydata` to be consistent with snippet above
- in the COPY mode box, the `open(..)` arrow should be two-directional to reflect writes can affect Local Path

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:88]:
```suggestion
2. **COPY** mode: The files are pre-fetched and cached on the local disk. Writes only affect the local copy and are not streamed back to the bucket.
```

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:87]: "The bucket" is more intuitive. "Object store" refers to the system (like S3).

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:62]: 3rd tab says `Create empty bucket` and reading it, my confusion is “Ok I can do that at first sky launch; but what about subsequent launch? the bucket has been created, can I still reuse the YAML without any changes"?

Do we mean "Create or reuse" here? This is an important use case, e.g., used in https://skypilot.readthedocs.io/en/latest/examples/spot-jobs.html#an-end-to-end-example.

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:31]: In all 3 tabs, can we be consistent with style of inline comments (2 spaces before `#`).

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:65]:
```suggestion
        To create an empty bucket, specify ``name`` (must be a unique name for the bucket).
```
or something like that. Point is to leave out "omit the source" since people may just click on this tab.

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:82]:
```suggestion
Storage modes
```

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:99]:
```suggestion
Picking a storage mode
```

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:110]: This is not very intuitive to me wearing an AI engineer's hat.

I'd have expected the first row should mention example use cases (Checkpoints (read-write access) vs. Datasets (read-only access)). Performance / Disk Size should come last.

**Comment by Michaelvll** [comment]
Just activate the hidden doc in our doc page for easier review: https://skypilot.readthedocs.io/en/storage-docs-rewrite/reference/storage.html
Will get to this soon : )

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:87]: We should mention "FUSE tools" somewhere on this page, maybe here.

**Review by concretevitamin — COMMENTED** [review]
Very clean! Quick look first.

File: docs/source/reference/storage.rst
Comment (Michaelvll) [inline docs/source/reference/storage.rst:37]: Does this mean we are not able to upload files to existing bucket? This might be unexpected, as in some of our users use case, they woud like to have all the data in a single bucket.

File: docs/source/reference/storage.rst
Comment (Michaelvll) [reply docs/source/reference/storage.rst:1]: +1

File: docs/source/reference/storage.rst
Comment (Michaelvll) [reply docs/source/reference/storage.rst:3]: +1 for this as well. Would it be possible to do `Cloud Buckets (Object Storage)`?

File: docs/source/reference/storage.rst
Comment (Michaelvll) [inline docs/source/reference/storage.rst:118]: For the disk size of the mount mode, we may have a footnote saying a smaller disk size than the objects to read may cause performance degradation?

File: docs/source/reference/storage.rst
Comment (Michaelvll) [inline docs/source/reference/storage.rst:150]: Should we mention the mounting an existing bucket as well in the common pattern as well?

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:3]: Changed to `Using Object Storages`

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:1]: Changed to `object storage` (to refer to architecture) and `bucket`(to refer to instances).

File: docs/source/reference/storage.rst
Comment (Michaelvll) [inline docs/source/reference/storage.rst:147]: I guess it would be nice to mention the ML specific use case in each of the following pattern in this section, e.g., `Storing and sharing outputs of tasks (model checkpoints)`, `Avoid re-uploading data on every run (uploading and sharing artifacts for multiple runs)`, `Copying data from storage before job starts (Dataset)` etc.

**Review by Michaelvll — COMMENTED** [review]
Thanks for revamping the doc for storage @romilbhardwaj! This page looks super nice. Left several comments about the ML specific use case clarification (similar as @concretevitamin 's comments).

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:110]: moved the "best for" column to the top

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:147]: good idea - I've added a **💡 Example use-case** under each pattern. does that help?

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:150]: Hmm that's covered in `Usage` at the top. Should we repeat it here?

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst]: good point - I've updated the title and added a visually obvious `note` box to mention if the bucket exists and was created by SkyPilot, it will be reused. Does that help?

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst]: added a note, does that help?

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst]: I've reworded the titles and added a visually distinct`note` - does that help?

**Comment by romilbhardwaj** [comment]
Thanks for the feedback @Michaelvll and @concretevitamin! Ready for another look.

File: docs/source/reference/storage.rst
Comment (Michaelvll) [reply docs/source/reference/storage.rst]: Ahh, I was wondering what will happen if the bucket already exists and was not created by SkyPilot. For example, there is a team writing files to the same bucket, and the bucket will only be `managed by SkyPilot` for one of the teammates.

File: docs/source/reference/storage.rst
Comment (Michaelvll) [reply docs/source/reference/storage.rst:162]: Ahh, I see. It might be good to avoid the duplication.

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:227]:
```suggestion
To manage buckets created by SkyPilot, the sky CLI provides two useful commands:
```

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:230]: Replace "storage objects" with "buckets" in this section?

File: docs/source/reference/yaml-spec.rst
Comment (concretevitamin) [inline docs/source/reference/yaml-spec.rst:270]:
```suggestion
        persistent: True  # Defaults to True; can be set to false to auto-delete bucket after cluster is downed. Optional.
```

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:162]:
```suggestion
Storing task outputs
```

File: docs/source/reference/storage.rst
Comment (concretevitamin) [reply docs/source/reference/storage.rst:162]: +1, I think it's good to add a "Reading datasets" (from an existing bucket) here. Most orgs have such setups.

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:185]:
```suggestion
Compared to directly using local paths in file_mounts, uploading to a bucket can
```

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:187]:
```suggestion
be faster since it is persistent.
```
Even if data is changing, this format is supported, right? (Like the note suggests below)

Comment (concretevitamin) [reply docs/source/reference/storage.rst:187]: Also, what's the reason this is faster vs. uploading to VMs? Is it persistent, or object storage having larger bandwidths than VMs.

**Review by concretevitamin — COMMENTED** [review]
Thanks @romilbhardwaj, did a pass.

**Comment by concretevitamin** [comment]
Also, should grep for `SkyPilot Storage` and rephrase. E.g., https://skypilot.readthedocs.io/en/storage-docs-rewrite/examples/syncing-code-artifacts.html

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:3]: `Cloud buckets` appears to be the more commonly used term:

image

Changed to just `Cloud Buckets`.

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:135]: Random writes will fail, appends IIRC fail on goofys (s3) and trigger a reupload on gcsfuse.

**Comment by romilbhardwaj** [comment]
Thanks @concretevitamin, ready for another look

File: docs/source/reference/storage.rst
Comment (Michaelvll) [reply docs/source/reference/storage.rst]: Just a kind bump for this @romilbhardwaj : )

Comment (romilbhardwaj) [reply docs/source/reference/storage.rst]: Ahh thanks for the ping @Michaelvll, this had slipped through the cracks.

If the bucket was created by a team member, other members will have to use the `source: s3://...` pattern. 

We disallow `name: external-bucket` pattern from mounting external buckets because e.g., if I want to create a bucket and use the name of a pre-existing public (or private) bucket, these semantics will silently mount the bucket and my task will fail when it tries to write the desired path (or worse yet, overwrite unintended files).

This behavior manifested in our smoke tests too (and is now fixed in #2804). Our `test_using_file_mounts_with_env_vars` was silently re-mounting `skypilot-temp-gcs-test` bucket which was shared in our GCS account, instead of creating a new bucket at each test instance.

File: docs/source/reference/storage.rst
Comment (concretevitamin) [inline docs/source/reference/storage.rst:180]:
```suggestion
Storing task outputs
~~~~~~~~~~~~~~~~~~~~
```

File: docs/source/examples/syncing-code-artifacts.rst
Comment (concretevitamin) [inline docs/source/examples/syncing-code-artifacts.rst:128]: L127: cloud buckets?

File: docs/source/reference/storage.rst
Comment (concretevitamin) [reply docs/source/reference/storage.rst]: Several options

1. "Use an existing bucket" / "Create and reuse a bucket" ("_and_" should avoid the conflict?)
1. "Use an external bucket" / "Create and reuse a bucket" 
2. "Use an existing bucket" / "Create a bucket"

Slight preference for 1 & 2, since I think the "reuse" part may be common to warrant being in the tabs; that said,  3 is ok with me too if @romilbhardwaj and @Michaelvll agree :)

**Review by concretevitamin — APPROVED** [review]
LGTM, some minor nits. Thanks @romilbhardwaj.

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst]: I'm fine with 2 or 3

Comment (Michaelvll) [reply docs/source/reference/storage.rst]: I would prefer 2 : )

File: docs/source/reference/storage.rst
Comment (Michaelvll) [reply docs/source/reference/storage.rst]: Thanks that makes sense! I found it a bit non-intuitive, for the following case, but I guess it is fine to leave it to the future.

If I am a member of a org, and created a bucket for with mount mode for storing some jupyter notebooks:
```
file_mounts:
  /dst:
    name: my-bucket
    store: s3
    mode: MOUNT
run: jupyter notebook /dst
```
The person then shares the YAML with another member in the org. Does this mean the other member running this yaml will fail and have to change it to:
```
file_mounts:
  /dst:
    source: s3://my-bucket
    mode: MOUNT
run: jupyter notebook /dst
```

File: docs/source/examples/syncing-code-artifacts.rst
Comment (Michaelvll) [inline docs/source/examples/syncing-code-artifacts.rst:127]: Seems we are using `cloud object storage` elsewhere, should we keep this aligned?

**Comment by romilbhardwaj** [comment]
Thanks for the reviews! Merging this soon - lmk if you have any other feedback @Michaelvll @concretevitamin

**Pull request merged**
